components/brave_wallet/renderer/js_cardano_provider.cc

```diff
@@ -0,0 +1,263 @@
+    for (const std::string& method :
+         {"getNetworkId", "getUsedAddresses", "getUnusedAddresses",
+          "getChangeAddress", "getRewardAddresses", "getUtxos", "getBalance",
+          "signTx", "signData", "submitTx", "getExtensions", "getCollateral"}) {
+      SetOwnPropertyWritable(context, wallet_api_object,
+                             gin::StringToV8(isolate, method), false);
+    }
+
+    std::ignore = resolver->Resolve(context, std::move(wallet_api_object));
+  } else {
+    std::ignore = resolver->Reject(context, result);
+  }
+}
+
+v8::Local<v8::Promise> JSCardanoProvider::IsEnabled(v8::Isolate* isolate) {
+  if (!EnsureConnected()) {
+    return v8::Local<v8::Promise>();
+  }
+
+  v8::Local<v8::Promise::Resolver> resolver_local;
+  if (!v8::Promise::Resolver::New(isolate->GetCurrentContext())
+           .ToLocal(&resolver_local)) {
+    return v8::Local<v8::Promise>();
+  }
+
+  auto global_context(
+      v8::Global<v8::Context>(isolate, isolate->GetCurrentContext()));
+  auto promise_resolver(
+      v8::Global<v8::Promise::Resolver>(isolate, resolver_local));
+
+  cardano_provider_->IsEnabled(base::BindOnce(
+      &JSCardanoProvider::OnIsEnableResponse, weak_ptr_factory_.GetWeakPtr(),
+      std::move(global_context), std::move(promise_resolver), isolate));
+
+  return resolver_local->GetPromise();
+}
+
+void JSCardanoProvider::OnIsEnableResponse(
+    v8::Global<v8::Context> global_context,
+    v8::Global<v8::Promise::Resolver> promise_resolver,
+    v8::Isolate* isolate,
+    const bool is_enabled) {
+  if (!render_frame()) {
+    return;
+  }
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Context> context = global_context.Get(isolate);
+  v8::Context::Scope context_scope(context);
+  v8::MicrotasksScope microtasks(isolate, context->GetMicrotaskQueue(),
+                                 v8::MicrotasksScope::kDoNotRunMicrotasks);
+
+  v8::Local<v8::Promise::Resolver> resolver = promise_resolver.Get(isolate);
+  std::ignore =
+      resolver->Resolve(context, v8::Boolean::New(isolate, is_enabled));
+}
+
+// static
+void JSCardanoProvider::Install(content::RenderFrame* render_frame) {
+  // TODO(https://github.com/brave/brave-browser/issues/46369): Add proxy object
+  // handler script
+  CHECK(render_frame);
+  v8::Isolate* isolate =
+      render_frame->GetWebFrame()->GetAgentGroupScheduler()->Isolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Context> context =
+      render_frame->GetWebFrame()->MainWorldScriptContext();
+  if (context.IsEmpty()) {
+    return;
+  }
+
+  v8::MicrotasksScope microtasks(isolate, context->GetMicrotaskQueue(),
+                                 v8::MicrotasksScope::kDoNotRunMicrotasks);
+  v8::Context::Scope context_scope(context);
+  v8::Local<v8::Object> global = context->Global();
+
+  v8::Local<v8::Value> cardano_root;
+  if (!global->Get(context, gin::StringToV8(isolate, kCardano))
+           .ToLocal(&cardano_root)) {
+    return;
```

Comment:
Are we expecting that this would return true instead of false when the context is undefined?

cc @cypt4 @supermassive just looking through this now for the sec review and came across it

Reply:
I am not sure what the case is, could you give more details please?
We have a [check](https://github.com/brave/brave-core/pull/29255/files/af7039cf249ae2253e0ff5c34cfc5e8ac0cef7e2#diff-94fc43aa3dcb6f822cca3fc609c6d082496cb25ee6421ea6d0b639580ff59bffR211) at the line 211 that context is not empty. Should we add more checks?